### DISCUSSION TITLE
Decoding the log file

### DISCUSSION THREAD

**Discussion by waltek001** [ERGO-Code/HiGHS#1736] [Q&A] [answered]
While the solver runs, it periodically updates with an improved solution. The corresponding line has a letter code that I presume indicates the methodology used to improve the solution. For example, here's a snippet showing R, but I've seen C and L as well:
```
Solving MIP model with:
   24768 rows
   322553 cols (2390 binary, 0 integer, 3 implied int., 320160 continuous)
   1631512 nonzeros

        Nodes      |    B&B Tree     |            Objective Bounds              |  Dynamic Constraints |       Work
     Proc. InQueue |  Leaves   Expl. | BestBound       BestSol              Gap |   Cuts   InLp Confl. | LpIters     Time

         0       0         0   0.00%   -inf            inf                  inf        0      0      0         0     7.6s
         0       0         0   0.00%   -57099573.48275 inf                  inf        0      0      2    145773   127.4s
 R       0       0         0   0.00%   -57086550.97073 -56477184.59953    1.08%       32      1      2    146131   139.1s
         0       0         0   0.00%   -57078622.67195 -56477184.59953    1.06%       64      8      2    146725   147.6s
```

Is there any documentation that describes what these codes mean?

Cheers,
Karl

**Comment by jajhall** [comment, marked as answer]
In short, no, but I got to the bottom of them recently when working on the MIP solver. and here's a key

- B => Branching; 
- C => Central rounding; 
- F => Feasibility pump; 
- H => Heuristic; 
- I => Initial; 
- L => Sub-MIP
- P => Empty MIP; 
- R => Randomized rounding; 
- S => Solve LP; 
- T => Evaluate node; 
- U => Unbounded;

To help decode this, but probably TMI, here are my detailed notes

bool HighsMipSolverData::addIncumbent(sol, solobj, source);

Use of source is only as the argument to HighsMipSolverData::printDisplayLine

When resolving the LP in HighsLpRelaxation

    1281: source is ‘S’

When handling special cases in mip/HighsMipSolverData.cpp

    79: Call from trySolution; source is passed in

bool HighsMipSolverData::trySolution computes the objective and row activities, returning false if the solution is not feasible, otherwise returns the truth of addIncumbent

trySolution is called from

        HighsLpRelaxation::run() if the node LP is unbounded; source is 'U'
        HighsPrimalHeuristics::solveSubMip when there is an optimal solution to a sub-MIP; source is 'L'

    469: Call from runSetup when mipsolver.numCol() == 0; source is ‘P’

    1361: Call from HighsMipSolverData::evaluateRootLp() when root node is optimal and has no fractional integers; source is ‘T’

When using primal heuristics (mip/HighsPrimalHeuristics.cpp)

    879: Call from tryRoundedPoint; source is passed in

bool HighsPrimalHeuristics::tryRoundedPoint is called from bool HighsPrimalHeuristics::linesearchRounding, which is called from

        HighsPrimalHeuristics::centralRounding() with source 'C'
        HighsPrimalHeuristics::feasibilityPump() with source 'F'

    1002: Call from randomizedRounding; source is ‘R’

    1111: Call from feasibilityPump; source is ‘F’

When performing vanilla searches (mip/HighsSearch.cpp)

    601: Call from selectBranchingCandidate; source is 'H' if inheuristic is true, otherwise 'B'

    734: Call from selectBranchingCandidate; source is 'H' if inheuristic is true, otherwise 'B'

    1064: Call from evaluateNode; source is 'H' if inheuristic is true, otherwise 'T'

**Reply by waltek001** [reply]
Perfect! Thank you!

K

**Reply by JunchengLiamLi** [reply]
@jajhall I am wondering if there are some references for how the primal heuristics in HiGHS work in details, e.g. academic papers or technical reports.

Many thanks,
Liam

**Reply by jajhall** [reply]
The primal heuristics are fairly standard. Perhaps @BenChampion could add some references

**Reply by BenChampion** [reply]
@JunchengLiamLi You could try this book which just came out earlier this month:

[Primal Heuristics in Integer Programming by Timo Berthold, Andrea Lodi, and Domenico Salvagnin](https://www.cambridge.org/core/books/primal-heuristics-in-integer-programming/FF179F5A69E7794BF0B737BEE0F92159)

It's an introductory survey with extensive references to the literature.

If you can't access that, you might look at Timo Berthold's thesis _Primal Heuristics for Mixed Integer Programs_ (2006). (The link I have is currently broken, but I'm sure you can find a copy somewhere.)

**Reply by JunchengLiamLi** [reply]
That is great. Thank you very much!
